tools/zap_logger.go

```diff
@@ -0,0 +1,95 @@
+package tools
+
+import (
+	"context"
+
+	"github.com/blendle/zapdriver"
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-logr/logr"
+	"github.com/go-logr/zapr"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+func NewZapLogger(logLevel zapcore.Level) (logr.Logger, zap.AtomicLevel, error) {
+	cfg := zapdriver.NewProductionConfig()
+	cfg.EncoderConfig.EncodeDuration = zapcore.StringDurationEncoder // we don't know why this is needed
+	cfg.EncoderConfig.EncodeTime = zapcore.RFC3339NanoTimeEncoder
+	cfg.Level.SetLevel(logLevel)
+	logger, err := cfg.Build()
+	if err != nil {
+		return logr.Logger{}, zap.AtomicLevel{}, err
+	}
+	return zapr.NewLogger(logger), cfg.Level, nil
+}
+
+func WatchLoggingConfig(ctx context.Context, watcher *fsnotify.Watcher, configFilepath string, logger logr.Logger, atomicLevel zap.AtomicLevel, getCfgLogLevelFunc func(string) (zapcore.Level, error)) error {
```

Comment:
This function is quite complex, codeclimate claims that it has 6 arguments and cognitive complexity of 36. Here are some refactoring ideas that might help:

* refactor the function to return a channel that a log level is being sent on an update. This would probably reduce the number of arguments (as `atomicLevel` and `getCfgLogLevelFunc` would be moved into the called).
* extract an utility that takes care about re-registering the watch upon path being deleted so that path handling is transparent to this function. This might be just a function that creates the watcher, setup a go routine to track path being deleted and return the watcher's `Event` channel

Reply:
Refactored this function. Code complexity is better, but still higher than codeclimate would like.